### ISSUE TITLE
In trainable segmentation, it is very slow to add annotations after a first segmentation has been performed

### ISSUE THREAD

**Issue by emmanuelle** [plotly/dash-sample-apps#456]
# App name

https://dash-gallery.plotly.host/dash-image-segmentation/

# Description of bug

After an initial segmentation has been performed, one often wishes to add new annotations to correct the model, but the loading component seems to be spinning forever and it is not possible to add new annotations. Unchecking "Show segmentation" does not seem to solve the problem.

Tagging you here @nicholas-esterer, I can also take a look after we decide together on priorities for the CZI apps. 

Ideally it should not take more time to add new annotations than before the segmentation has been computed.

![segmentation-bug](https://user-images.githubusercontent.com/263366/86804424-33262300-c077-11ea-83ea-f5da6e56f38e.gif)

**Comment by nicholas-esterer** [comment]
It's true adding additional shapes makes annotating take a little bit longer. But running locally every time I add a shape it keeps working, and I can turn on/off show segmentations. I guess what I'm trying to say is it seems to be something to do with when it is deployed and so I'd like to discuss how we might go about diagnosing this?

**Comment by emmanuelle** [comment]
Ah interesting, thanks for raising this point. I confirm that it runs smoothly locally. About diagnosis, is your question:
- how can we diagnose where the problem comes from now that it exists? If it's fast locally and slow on the network, my guess is that the bottleneck is the transfer of data? The only way I know is to read the callback code and to see what changes in terms of Input and Output after a first segmentation.
- or, how can we develop so that we work always with a deployed version so that we can catch such problems early enough and not at the end? For this, the solution can be do deploy very frequently to playground (probably by pushing commits to a branch here on dash-sample-apps), or to work with Dash Enterprise workspaces so that the development is directly on the server and you always check a deployed version.

**Comment by nicholas-esterer** [comment]
I guess it is both, but I only have insight for the first kind of problem, haha (I never thought about DE workspaces though, that might help, although I'd prefer to keep using the tools I use normally if possible :) ).
I can think of a couple of places where we can save bandwidth
 - right now the image is uploaded to the client every time the annotations change. For the 3D app I figured out how to make that unnecessary by taking the "background" image from a store and just uploading the segments. So we could use a similar approach in this app. The background image would be loaded once upon initial page load.
 - the training doesn't use any prior information when you add new points (it just starts the training from the beginning). I remember you saying that it could be possible to use the old training and just add the new shapes, and continue training?

**Comment by emmanuelle** [comment]
>     * the training doesn't use any prior information when you add new points (it just starts the training from the beginning). I remember you saying that it could be possible to use the old training and just add the new shapes, and continue training?

We could use the `warm_start` parameter of https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html but I'm not sure this is the right use of `warm_start` (see https://scikit-learn.org/stable/glossary.html#term-warm-start). It's worth trying though!

**Comment by emmanuelle** [comment]
Also, are features recomputed each time? The result could be cached.

**Comment by nicholas-esterer** [comment]
Right now there's a hash of the fig.shapes made and if these haven't changed, nothing is recomputed.
However if shapes have changed, then yes as long as the "Show segmentation" checkbox is active, the new segmentation is computed and shown. As it is right now, I'm not sure how the result could be cached, unless we set up some kind of mechanism to just transmit the diff of of the segmentation rather than the the whole thing.
That being said, I really think the delay is just in the classifier, because it is non-trivial code, so using some kind of `warm_start` might help with this. Another thing I've experienced in the past is a module import taking a long time to import (in that case it was computing filter impulse responses or something): if the server boots another worker to handle a request, this might be served more slowly because the script needs to import everything again, which might take time.
